source/ncconvert/utility/Image.cpp

```diff
@@ -0,0 +1,168 @@
+    std::memcpy(out.data(), image.get(), numBytes);
+    return out;
+}
+
+auto GetPaddedDimension(int dimension) -> int
+{
+    return (dimension + 3) & ~3;
+}
+} // anonymous namespace
+
+namespace nc::convert
+{
+auto MakeRawImage(int width,
+                  int height,
+                  const unsigned char fillRGBA[4]) -> RawImage
+{
+    const auto numBytes = width * height * 4;
+    auto bytes = (unsigned char*)std::malloc(numBytes);
+    for (auto i = 0; i < numBytes; i += 4)
+    {
+        std::memcpy(bytes + i, fillRGBA, 4);
+    }
+
+    return RawImage{bytes};
+}
+
+Image::Image(const std::filesystem::path& path)
+{
+    const auto pathString = path.string(); // because path.c_str() is wchar* on Windows
+    auto nChannels = int32_t{};
+    auto ptr = stbi_load(pathString.c_str(), &m_width, &m_height, &nChannels, STBI_rgb_alpha);
+    if(!ptr)
+    {
+        throw NcError("Failed to load texture file: ", pathString);
+    }
+
+    m_data = RawImage{ptr};
+}
+
+auto Image::MakeTextureSubResource() const -> asset::TextureSubResource
+{
+    return asset::TextureSubResource{
+        .width = static_cast<uint32_t>(m_width),
+        .height = static_cast<uint32_t>(m_height),
+        .pixelData = ToVector(m_data, GetSizeInBytes())
+    };
+}
+
+void Image::ResizePadded()
+{
+    if (!RequiresCompressionPadding())
+    {
+        return;
+    }
+
+    const auto paddedWidth = GetPaddedDimension(m_width);
+    const auto paddedHeight = GetPaddedDimension(m_height);
+    auto paddedImage = MakeRawImage(paddedWidth, paddedHeight, paddingRGBA);
+
+    auto dst = paddedImage.get();
+    const auto src = m_data.get();
+    const auto dstScanLineSize = paddedWidth * numChannels;
+    const auto srcScanLineSize = m_width * numChannels;
+    for (auto scanLine = 0; scanLine < m_height; ++scanLine)
+    {
+        const auto dstLine = dst + scanLine * dstScanLineSize;
+        const auto srcLine = src + scanLine * srcScanLineSize;
+        std::memcpy(dstLine, srcLine, srcScanLineSize);
+    }
+
+    m_data = std::move(paddedImage);
+    m_width = paddedWidth;
+    m_height = paddedHeight;
+}
+
+auto Image::Compress(asset::TextureFormat format) const -> asset::TextureSubResource
+{
+    NC_ASSERT(
+        IsCompressedTextureFormat(format) && !RequiresCompressionPadding(),
+        fmt::format(
+            "Invalid parameters for compression. Resize image or use an appropriate format.\n"
+            "\tTextureFormat '{}', Width '{}', Height '{}'",
+            std::to_underlying(format),
+            m_width,
+            m_height
+        )
+    );
+
+    constexpr auto blockSize = 16;
+    const auto useAlpha = TextureFormatHasAlpha(format);
+    const auto numBlocks = (m_width / 4) * (m_height / 4);
+    auto compressedData = std::vector<unsigned char>(numBlocks * blockSize);
+    unsigned char block[blockSize * 4];
+
+    // compress in 4x4 RGBA blocks
+    const auto src = m_data.get();
+    for (auto y = 0; y < m_height; y += 4)
```

Comment:
This might be a case for more specific index variable names

Reply:
fixed